### ISSUE TITLE
% can be greater than 100

### ISSUE THREAD

**Issue by oscardssmith** [vpelletier/pprofile#24]
I got the following line in output
`(call)|     13592|      226.612|    0.0166725|1277.08%|# test.py:1 composite_candidates`

Here is the relevent code

    def composite_candidates(primes, limit, factors=(1,), num=2, candidates=None, idx=0):
        if candidates == None:
            candidates = {1:1, 2:2, 4:3, 36:9}
        if num < limit:
            # current exponent must be equal or less than previous one
            if len(factors) == 1 or factors[-2] > factors[-1]:
                # we either stay with current prime and increase exponent
                composite_candidates(primes, limit, 
                                     factors[:-1] + (factors[-1] + 1,), 
                                     num * primes[idx], candidates, idx)
            # or move to next prime
            idx += 1
            factors = factors + (1,)
            num *= primes[idx]
            composite_candidates(primes, limit, factors, num, candidates, idx)
            
            # and add the candidate if it might be right
            if num < limit:
                if primes[idx].bit_length()-1 <= factors[0] <= 2 * primes[idx+1].bit_length():
                    divisors = num_factors(factors)
                    #candidates[num] = divisors
                    candidates[num ]= divisors
            
        return candidates
        
    def f(limit):
        primes = prime_sieve(2*limit.bit_length() + 5)
        candidates = composite_candidates(primes, limit)
    f(2**60)

**Comment by vpelletier** [comment]
Do you use multithreading, with (at least) 13 threads ? This could be a reason.

**Comment by oscardssmith** [comment]
No threading. It's all sequential.

**Comment by vpelletier** [comment]
Reproduced, this is about recursion, indeed `(call)` lines do get above 100%.

**Comment by oscardssmith** [comment]
Do you have any idea how to fix this? I would be happy to submit a PR, but I don't know where to start.

**Comment by oscardssmith** [comment]
I think I've figured out part of the problem. I believe that a recursive call adds the time taken both for the hit, and also for the return. Since it is the same function, this double counts. That still doesn't explain how the result can be higher than 200% though...

**Comment by vpelletier** [comment]
The problem is that the time from call to return is added per call site and callable (ie, if a single statement calls more than one callable, two "call" lines are shown with their respective timings).

If there is a 2-level, single-branch recursion timing becomes (taking the number lines as a very crude timing):
```
+> foo()
|  +> foo()
|  +> return : t=2
+-> return : t=4
```
Result is that `foo` cost a total time of 6 lines, while overall script ran for only 4, so 150% call time.

Likewise for multi-branches:
```
+> foo()
|  +> foo()
|  +> return : t=2
|  +> foo()
|  +> return : t=2
+-> return : t=6
```
Total is now 10 with total time of 6, or 166%.

And likewise for multi-call-layers:
```
+> foo()
|  +> foo()
|  |  +> foo()
|  |  +> return : t=2
|  +> return : t=4
+-> return : t=6
```
Total is now 12, or 200%.

Of course, on a real script with realistic timing, call & return take close to 0 and function payload most of the time, so a 2-level recursion will tend to 200% on its call-site, 3-level to 300% and so on. And multi-branches make it exponential.

I believe this is the cause of the issue.

Now, about a solution... I could not figure one yet.

A difficulty is that recursion is not necessarily only a callable calling itself, but it can go through another one:
```
+> foo()
|  +> bar()
|  |  +> foo()
|  |  +> return : t=2
|  +> return : t=4
+-> return : t=6
```

Here, I would like to discount 2 out of the 4 from `bar` when returning to topmost `foo`, because these 2 belong to `foo` already. So I need pprofile to remember it is already in topmost `foo` when it enters (or returns, actually) from any inner `foo`. So this feels like refcounting (+1 for a call, -1 for a return) in a dict (key being callee and call site).

But it gets worse: here, I want the line calling `foo` from `bar` to show a call time of "2", so I still want to count that as call time, but that time must be ignored in the total for the recursive call site. So the cost must also be discounted from `bar` total call time, *but* only when returning to `foo`. So now pprofile also needs to care about all the calls which happened inside the stack it is returning from, and not just the stack above it. Which looks like a potential memory hell.

This is about where I started wandering off, hoping for a solution to eventually pop up when I stopped thinking about it. And it did not hit me.

**Comment by oscardssmith** [comment]
I think I might have a solution, but it is really messy. Instead of recording the time function calls take, you could instead build a tree of the function calls for a program that would store the name, time, and children. Then you would get the total time after the program was run by traversing the tree from the bottom, and making each function only report adding time for each non self function.

**Comment by vpelletier** [comment]
It needs to take into account each call site, not just each callee. I'm not sure how it will behave (memory-wise) on large programs, where many pieces get called from many places, but I'm curious.

**Comment by scotthuang1989** [comment]
I think multi-thread program also have same problem.

**Comment by oscardssmith** [comment]
I'm sure they do, but the single threaded case will almost certainly be easier to fix, so it makes sense to start there.

**Comment by vpelletier** [comment]
I think I found how to efficiently implement the fix for this issue: e938c31c85ef40e0c0e050bb5522065db462dac8 . Please give it a try.

This does not change how threading is handled. I tend to think >100% in a threaded case is representative of reality: this means that "stuff" could happen in parallel, so that more than 100% of "work" could be done between given wall-clock start and stop times. Maybe "work" is waiting for a certain time to pass or waiting for file descriptor activity (so not really CPU work), but that is consistent with how such calls are accounted in single threaded code.

An alternative for threading would be to provide separate results for separate threads. I'm not sure how usable the output would be.

**Comment by vpelletier** [comment]
FWIW, I committed the [key by caller and callee](fdb7e5fed0f99d7c362ed37dca3db606ae29e714), if you want to give it a try.

**Comment by vpelletier** [comment]
Closing as fixed in 2.0.0, which I released a few hours ago.

**Comment by oscardssmith** [comment]
Can confirm. This works. Thanks
